sporkfish/searcher/minimax.py

```diff
@@ -171,41 +187,114 @@ def captured_piece(board: Board, move: chess.Move) -> chess.PieceType:
 
         return alpha
 
+    def _futility_pruning(
+        self, board: Board, depth: int, capture: bool, move: chess.Move, alpha: float
+    ) -> bool:
+        """
+        Implements futility pruning.
+
+        If the evaluation of the current position, when extended by a margin, falls below the minimum score
+        we can guarantee, then it's not worthwhile to continue the search.
+        However, it's important to note that we still need to consider tactical possibilities due to captures,
+        promotions, and checks.
+
+        :param board: The current board state.
+        :type board: chess.Board
+        :param depth: The current depth in the search tree.
+        :type depth: int
+        :param capture: Indicates if the previous move was a capture.
+        :type capture: bool
+        :param move: The move that was made.
+        :type move: chess.Move
+        :param alpha: The current best score for the maximizing player.
+        :type alpha: float
+        :return: True if the position can be pruned due to futility margin checks, False otherwise.
+        :rtype: bool
+        """
+        if depth <= 3 and not capture and not move.promotion and not board.is_check():
+            # TODO: consider using different futility margins
+            # Half a pawn margin is very aggressive
+            if (
+                self._evaluator.evaluate(board)
+                + depth * self.evaluator.MG_PIECE_VALUES[chess.PAWN] // 2
```

Comment:
I tried to encourage as much pruning as possible basically. The call to evaluator is so expensive that if we don't manage prune we add a fairly unreasonable amount of time (as you can see in the results).

Re MG vs EG, detecting the phase is theoretically possible with our current design but practically unfeasible. Calculating game phase requires checking the board for remaining pieces, which is costly. We do this in evaluator and can propagate this back up here, but I don't see much benefit to overtune values we already don't have much understanding of. Moreover we have EG table bases in the work by @michelle-lai-swl so this will be more used in MG I think.

Reply:
Re MG vs EG - all clear - tyty

Sorry I am still not clear why we are using pawn here as margin though? Reading the [reference link](https://www.chessprogramming.org/Futility_Pruning#Extended_Futility_Pruning) you put on READ, it says "If at depth 1 the margin does not exceed the value of a minor piece, at depth 2 it should be more like the value of a rook."

Reply:
I don't have much rationale beyond wanting to prune as much as possible. Perhaps it's not good but otherwise from testing the cost is too great without much benefit. 

The engine cpw-engine uses (100,200,400) for depth(i) in (1,2,3) which differs from the reference in the wiki, so it doesn't seem like there's much of a fixed guideline on what margin is best

Reply:
It is a bit risky to naively "prune as much as possible" though? As we know, futility pruning is known to not necessarily return the theoretically best move - i.e. to me, the margin size is a trade off between accuracy and speed. 

This means the choice of margin size is to be optimised. Is my understanding correct?

Don't want to make a huge fuss about it for now, given that it is not turned on. I think we should have the functionality in first, and we can do testing later. Most important thing is that it is not breaking other things, and is doing what we think it should be doing.

Reply:
Yes - later on we need to a check to make sure we don't prune the PV move (from previous depth) as well.